codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/generators/EndpointParamsInterceptorGenerator.kt

```diff
@@ -111,7 +111,7 @@ class EndpointParamsInterceptorGenerator(
             )
             withBlockTemplate(
                 "let endpoint_prefix = ",
-                ".map_err(#{InterceptorError}::read_before_execution)?;",
+                """.map_err(|err| format!("endpoint prefix could not be built: {err:?}"))?;""",
```

Comment:
I'm starting to think we need some kind of `ErrorMessage` struct (perhaps there's a better name for it) that is just a string message and a `Box<dyn Error>` source. We shouldn't be string formatting errors.

This could be something like:
```rust
.map_err(|err| ErrorMessage::new("endpoint prefix could not be built", err))?
```

Essentially we just need a convenient way to attach additional context to an error. Or maybe better, could do something inverse with traits:

```rust
.map_err(|err| "endpoint prefix could not be built".source(err))?
```

Reply:
I like that idea. Added a convenience type `ContextAttachedError` in 51026e9.